core/c11_atomic.c

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2015 Eistec AB
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup         core_util
+ * @{
+ *
+ * @file
+ *
+ * @brief Implementation of C11 atomic operations if GCC does not provide an
+ * implementation.
+ *
+ * GCC with -mcpu=cortex-m3 and cortex-m4 generate LDREX/STREX instructions
+ * instead of library calls. There is however currently (2015-05-29) no
+ * implementation for Cortex-M0, since it lacks the lock-free atomic operations
+ * found in the M3 and M4 cores.
+ *
+ * @note Other CPUs (e.g. msp430, avr) might need this too, but current MSP430
+ * GCC in Ubuntu/Debian is stuck at version 4.6 which does not provide C11
+ * language support which makes it difficult to actually make use of this on
+ * that platform.
+ *
+ * @note This implementation completely ignores the memory model parameter
+ *
+ * @see https://gcc.gnu.org/wiki/Atomic/GCCMM/LIbrary
+ *
+ * @author Joakim Gebart <joakim.gebart@eistec.se>
+ */
+
+/* no need to #ifdef these because GCC will only generate calls to the helper
+ * functions if hardware support is missing */
+#if 1
+
+#include <stdint.h>
+#include "irq.h"
+
+/* GCC documentation refers to the types as I1, I2, I4, I8, I16 */
+typedef uint8_t  I1;
+typedef uint16_t I2;
+typedef uint32_t I4;
+typedef uint64_t I8;
+/* typedef __uint128_t I16; */ /* No 128 bit integer support yet */
+
+/**
+ * @brief This is a macro that defines a function named __atomic_fetch_<em>opname</em>_<em>n</em>
+ *
+ * \param opname    operator name that will be used in the function name
+ * \param op        actual C language operator
+ * \param n         width of the data, in bytes
+ * \param prefixop  optional prefix unary operator (use ~ for inverting, NAND, NOR etc)
+ */
+#define TEMPLATE_ATOMIC_FETCH_OP_N(opname, op, n, prefixop) \
+    I##n __atomic_fetch_##opname##_##n (volatile void *ptr, I##n val, int memmodel) \
+    { \
+        unsigned int mask = disableIRQ();     \
+        (void)memmodel;                       \
+        I##n tmp = *((I##n*)ptr);             \
+        *((I##n*)ptr) = prefixop(tmp op val); \
+        restoreIRQ(mask);                     \
+        return tmp;                           \
+    }
+
+/* Template instantiations below */
+TEMPLATE_ATOMIC_FETCH_OP_N( add, +=, 1,  ) /* __atomic_fetch_add_1 */
```

Comment:
`__atomic_fetch_add()` and friends should return the old value, shan't they? Then `op` should be `+` instead of `+=`.